proposals/improved-interpolated-strings.md

```diff
@@ -0,0 +1,436 @@
+* Pass parameters to the builder marked with a specific attribute, a la `EnumeratorCancellation` support. This would need rules about whether we pass the receiver (maybe if the method is marked
+we pass the receiver, and we don't in the general case?), and what we do if parameters _after_ the string parameter are annotated, but it seems like a potential option.
+
+Some compromise is likely needed here, but either direction has complications. Some scenarios that would be affected by this is the `Utf8Formatter` below, or existing api patterns that have
+an `IFormatProvider` as the first argument.
+
+### `await` usage in interpolation holes
+
+Because `$"{await A()}"` is a valid expression today, we need to rationalize how interpolation holes with await. We could solve this with a few rules:
+
+1. If an interpolated string used as a `string`, `IFormattable`, or `FormattableString` has an `await` in an interpolation hole, fall back to old-style formatter.
+2. If an interpolated string is subject to an _implicit\_string\_builder\_conversion_ and _applicable\_interpolated\_string\_builder\_type_ is a `ref struct`, `await` is not allowed to be used
+in the format holes.
+
+Fundamentally, this desugaring could use a ref struct in an async method as long as we guarantee that the `ref struct` will not need to be saved to the heap, which should be possible if we forbid
+`await`s in the interpolation holes.
+
+Alternatively, we could simply make all builder types non-ref structs, including the framework builder for interpolated strings. This would, however, preclude us from someday recognizing a `Span`
+version that does not need to allocate any scratch space at all.
+
+### Builders as ref parameters
+
+Some builders might want to be passed as ref parameters (either `in` or `ref`). Should we allow either? And if so, what will a `ref` builder look like? `ref $""` is confusing, as you're not actually
+passing the string by ref, you're passing the builder that is created from the ref by ref, and has similar potential issues with async methods.
+
+## Other use cases
+
+### `TryFormat` on `Span` receivers
+
+The BCL has a number of helper methods that and usages of `ValueStringBuilder` that attempt to format a given string into a `Span`, and instead of moving to the heap if needed, give up if
+the `Span` isn't big enough to hold the resulting text. With this proposal, it would be possible to support these cases by defining an extension method that looks like this:
+
+```cs
+public static class MemoryExtensions
+{
+    public static bool TryWrite(this Span<char> span, SpanInterpolatedStringBuilder builder, out int charsWritten)
+    {
+        charsWritten = builder._count;
+        return builder._success;
+    }
+ 
+}
+ 
+public ref struct SpanInterpolatedStringBuilder
+{
+    public static void GetInterpolatedStringBuilder(int baseLength, int formatHoleCount, Span<char> span, out SpanInterpolatedStringBuilder builder)
```

Comment:
I wonder whether GetInterpolatedStringBuilder should return a bool.

The TryFormat methods can all assume that the compiler generated code won't call them if a previous TryFormat returned false.  However, in a case like this builder, all of the TryFormats need to be prepared for the possibility that the initial configuration was flawed, e.g. that the baseLength > span.Length, and so they all need to have upfront `if (!_success) return false;` checks.  If GetInterpolatedStringBuilder returned a bool, and the generated code could then be more like:
```C#
_ = GetInterpolatedStringBuilder(..., out var builder) &&
    builder.TryFormat(...) &&
    builder.TryFormat(...);
```
then the individual TryFormat methods would need to check _success.